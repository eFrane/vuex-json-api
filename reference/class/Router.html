<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Json:Api for Vuex</title>
    <meta name="generator" content="VuePress 1.9.10">
    
    <meta name="description" content="Automagic Json:Api integration for Vuex">
    
    <link rel="preload" href="/assets/css/0.styles.769add7b.css" as="style"><link rel="preload" href="/assets/js/app.9efa2014.js" as="script"><link rel="preload" href="/assets/js/2.fb51603b.js" as="script"><link rel="preload" href="/assets/js/1.2ea433ec.js" as="script"><link rel="preload" href="/assets/js/36.f6a75ce8.js" as="script"><link rel="prefetch" href="/assets/js/10.5f385d4d.js"><link rel="prefetch" href="/assets/js/11.ba710dea.js"><link rel="prefetch" href="/assets/js/12.46048386.js"><link rel="prefetch" href="/assets/js/13.357634dc.js"><link rel="prefetch" href="/assets/js/14.65769da9.js"><link rel="prefetch" href="/assets/js/15.b2acdef1.js"><link rel="prefetch" href="/assets/js/16.81bd89c9.js"><link rel="prefetch" href="/assets/js/17.459cf0b5.js"><link rel="prefetch" href="/assets/js/18.285b93d5.js"><link rel="prefetch" href="/assets/js/19.2b3d35ee.js"><link rel="prefetch" href="/assets/js/20.b6603841.js"><link rel="prefetch" href="/assets/js/21.ea8b26f4.js"><link rel="prefetch" href="/assets/js/22.dcb39a39.js"><link rel="prefetch" href="/assets/js/23.fa408d10.js"><link rel="prefetch" href="/assets/js/24.69e34714.js"><link rel="prefetch" href="/assets/js/25.3ca9d6b3.js"><link rel="prefetch" href="/assets/js/26.62a29a7a.js"><link rel="prefetch" href="/assets/js/27.87f07c1c.js"><link rel="prefetch" href="/assets/js/28.95b71fce.js"><link rel="prefetch" href="/assets/js/29.2626e755.js"><link rel="prefetch" href="/assets/js/3.bee1657c.js"><link rel="prefetch" href="/assets/js/30.1567c6aa.js"><link rel="prefetch" href="/assets/js/31.2b529aa9.js"><link rel="prefetch" href="/assets/js/32.afee7aa5.js"><link rel="prefetch" href="/assets/js/33.a52dd89e.js"><link rel="prefetch" href="/assets/js/34.fba2356b.js"><link rel="prefetch" href="/assets/js/35.1e223504.js"><link rel="prefetch" href="/assets/js/37.c42e0dd6.js"><link rel="prefetch" href="/assets/js/38.ba646047.js"><link rel="prefetch" href="/assets/js/39.4f6dc3a8.js"><link rel="prefetch" href="/assets/js/4.52dc2f05.js"><link rel="prefetch" href="/assets/js/40.9cf469f9.js"><link rel="prefetch" href="/assets/js/41.e2982e59.js"><link rel="prefetch" href="/assets/js/42.0e52e9a2.js"><link rel="prefetch" href="/assets/js/43.2250b9a9.js"><link rel="prefetch" href="/assets/js/44.511c4928.js"><link rel="prefetch" href="/assets/js/45.586c49c5.js"><link rel="prefetch" href="/assets/js/46.ecd1fc23.js"><link rel="prefetch" href="/assets/js/47.1554335c.js"><link rel="prefetch" href="/assets/js/48.77cb5f20.js"><link rel="prefetch" href="/assets/js/49.b7c951c8.js"><link rel="prefetch" href="/assets/js/5.18620486.js"><link rel="prefetch" href="/assets/js/50.6beb7ee7.js"><link rel="prefetch" href="/assets/js/51.1f9a772e.js"><link rel="prefetch" href="/assets/js/52.3fae7d23.js"><link rel="prefetch" href="/assets/js/53.0a5a39dd.js"><link rel="prefetch" href="/assets/js/54.7afc0b6c.js"><link rel="prefetch" href="/assets/js/55.9bb55334.js"><link rel="prefetch" href="/assets/js/56.907e5f15.js"><link rel="prefetch" href="/assets/js/57.a2e0b921.js"><link rel="prefetch" href="/assets/js/58.edbfbb04.js"><link rel="prefetch" href="/assets/js/59.c26a30a4.js"><link rel="prefetch" href="/assets/js/6.891bb6ed.js"><link rel="prefetch" href="/assets/js/60.9765273a.js"><link rel="prefetch" href="/assets/js/61.e1898988.js"><link rel="prefetch" href="/assets/js/62.45996bc1.js"><link rel="prefetch" href="/assets/js/63.b2c487ae.js"><link rel="prefetch" href="/assets/js/64.2790a4b5.js"><link rel="prefetch" href="/assets/js/65.052dd7c1.js"><link rel="prefetch" href="/assets/js/66.9fee64b0.js"><link rel="prefetch" href="/assets/js/67.007a4b9c.js"><link rel="prefetch" href="/assets/js/68.68ea8eb5.js"><link rel="prefetch" href="/assets/js/69.c9e0043e.js"><link rel="prefetch" href="/assets/js/7.243f6fed.js"><link rel="prefetch" href="/assets/js/70.ed0081f6.js"><link rel="prefetch" href="/assets/js/71.b8809075.js"><link rel="prefetch" href="/assets/js/72.15bb2736.js"><link rel="prefetch" href="/assets/js/73.2582b43e.js"><link rel="prefetch" href="/assets/js/74.6da308f8.js"><link rel="prefetch" href="/assets/js/vendors~docsearch.cdc87792.js">
    <link rel="stylesheet" href="/assets/css/0.styles.769add7b.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/" class="home-link router-link-active"><!----> <span class="site-name">Json:Api for Vuex</span></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <nav class="nav-links can-hide"><div class="nav-item"><a href="/" class="nav-link">
  Home
</a></div><div class="nav-item"><a href="/development/" class="nav-link">
  Development
</a></div><div class="nav-item"><a href="/reference/" class="nav-link router-link-active">
  Reference
</a></div><div class="nav-item"><a href="/usage/" class="nav-link">
  Usage
</a></div> <a href="https://github.com/eFrane/vuex-json-api" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></nav></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><nav class="nav-links"><div class="nav-item"><a href="/" class="nav-link">
  Home
</a></div><div class="nav-item"><a href="/development/" class="nav-link">
  Development
</a></div><div class="nav-item"><a href="/reference/" class="nav-link router-link-active">
  Reference
</a></div><div class="nav-item"><a href="/usage/" class="nav-link">
  Usage
</a></div> <a href="https://github.com/eFrane/vuex-json-api" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></nav>  <ul class="sidebar-links"><li><section class="sidebar-group depth-0"><p class="sidebar-heading open"><span></span> <!----></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/reference/class/Router.html#router" class="sidebar-link">Router</a><ul class="sidebar-sub-headers"><li class="sidebar-sub-header"><a href="/reference/class/Router.html#new-router" class="sidebar-link">new Router()</a></li><li class="sidebar-sub-header"><a href="/reference/class/Router.html#router-getroutes-⇒-array-route" class="sidebar-link">router.getRoutes() ⇒</a></li><li class="sidebar-sub-header"><a href="/reference/class/Router.html#router-addroute-route" class="sidebar-link">router.addRoute(route)</a></li></ul></li></ul></section></li></ul> </aside> <main class="page"> <div class="theme-default-content content__default"><p><a name="Router"></a></p> <h2 id="router"><a href="#router" class="header-anchor">#</a> Router</h2> <p><strong>Kind</strong>: global class</p> <ul><li><a href="#Router">Router</a> <ul><li><a href="#new_Router_new">new Router()</a></li> <li><a href="#Router+getRoutes">.getRoutes()</a> ⇒ <code>Array.&lt;Route&gt;</code></li> <li><a href="#Router+addRoute">.addRoute(route)</a></li></ul></li></ul> <p><a name="new_Router_new"></a></p> <h3 id="new-router"><a href="#new-router" class="header-anchor">#</a> new Router()</h3> <p>Basic router implementation for the ResourcefulApi.</p> <p>Automagically creating api bound modules builds on
an understanding of the available routes. To
easily instantiate a Store bound to an endpoint,
route information for that endpoint must be provided.</p> <p>Since every endpoint is implemented differently and
the choice where this route information comes from
should be left to the endpoint developer, this
library only assumes that route loading can be a
asynchronous process which eventually returns and
has a set of <code>Route</code> objects in <code>this.routes</code>,
keyed by their module and methods. To ensure the latter,
actually adding <code>Route</code>s should be done via <code>addRoute</code>.</p> <p><a name="Router+getRoutes"></a></p> <h3 id="router-getroutes-⇒-array-route"><a href="#router-getroutes-⇒-array-route" class="header-anchor">#</a> router.getRoutes() ⇒ <code>Array.&lt;Route&gt;</code></h3> <p><strong>Kind</strong>: instance method of <a href="#Router"><code>Router</code></a><br> <a name="Router+addRoute"></a></p> <h3 id="router-addroute-route"><a href="#router-addroute-route" class="header-anchor">#</a> router.addRoute(route)</h3> <p><strong>Kind</strong>: instance method of <a href="#Router"><code>Router</code></a></p> <table><thead><tr><th>Param</th> <th>Type</th></tr></thead> <tbody><tr><td>route</td> <td><code>Route</code></td></tr></tbody></table></div> <footer class="page-edit"><!----> <!----></footer> <div class="page-nav"><p class="inner"><span class="prev">
      ←
      <a href="/reference/" class="prev router-link-active">
        Reference (JS)
      </a></span> <!----></p></div> </main></div><div class="global-ui"></div></div>
    <script src="/assets/js/app.9efa2014.js" defer></script><script src="/assets/js/2.fb51603b.js" defer></script><script src="/assets/js/1.2ea433ec.js" defer></script><script src="/assets/js/36.f6a75ce8.js" defer></script>
  </body>
</html>
