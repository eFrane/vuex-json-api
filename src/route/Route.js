export class Route {
  /**
   *
   * @param {String} module
   * @param {String} action
   * @param {String} url
   * @param {Array} parameters
   */
  constructor (module, action, url, parameters) {
    if (typeof module !== 'string') {
      throw new Error('Module name must be string')
    }

    if (!this.isValidAction(action)) {
      throw new Error('Action must be valid resourceful verb')
    }

    if (typeof url !== 'string') {
      throw new Error('URL must be string')
    }

    if (parameters.constructor !== Array) {
      throw new Error('Parameters must be array')
    }

    this.module = module
    this.action = action.toLowerCase()
    this.url = url
    this.parameters = parameters
  }

  get supportedResourcefulVerbs () {
    return [
      'list',
      'get',
      'create',
      'replace',
      'update',
      'delete'
    ]
  }

  /**
   * Prepare an url, replacing provided parameters
   *
   * @param parameters
   * @returns {string}
   */
  prepare (parameters) {
    let url = this.url

    for (const param in parameters) {
      if (Object.prototype.hasOwnProperty.call(parameters, param) &&
        this.hasParameter(param)) {
        url = url.replace('{' + param + '}', parameters[param])
        delete parameters[param]
      }
    }

    return url
  }

  /**
   * Check if a parameter is allowed
   *
   * @param parameter
   * @returns {boolean}
   */
  hasParameter (parameter) {
    return this.parameters.filter((checkParam) => {
      return parameter === checkParam
    }).length > 0
  }

  isAbsolute () {
    return this.url.substring(0, 4) === 'http'
  }

  /**
   *
   * @param {String} action
   */
  isValidAction (action) {
    return typeof action === 'string' && this.supportedResourcefulVerbs.indexOf(action.toLowerCase()) > -1
  }
}
