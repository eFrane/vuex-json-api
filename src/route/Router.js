/**
 * Basic router implementation for the ResourcefulApi.
 *
 * Automagically creating api bound modules builds on
 * an understanding of the available routes. To
 * easily instantiate a Store bound to an endpoint,
 * route information for that endpoint must be provided.
 *
 * Since every endpoint is implemented differently and
 * the choice where this route information comes from
 * should be left to the endpoint developer, this
 * library only assumes that route loading is usually an
 * asynchronous process which eventually returns and
 * has a set of keyed-by-name `Route` objects in
 * `this.routes`.
 */
export class Router {
  constructor () {
    this.routes = {}
  }

  getRoutes () {
    return this.routes
  }

  async updateRoutes () {
    return this
  }

  /**
   *
   * @param {Route} route
   */
  addRoute (route) {
    if (!Object.prototype.hasOwnProperty.call(this.routes, route.module)) {
      this.routes[route.module] = {}
    }

    this.routes[route.module][route.action] = route
  }
}
