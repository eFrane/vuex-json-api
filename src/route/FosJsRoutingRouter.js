import { Route } from './Route'
import { Router } from './Router'

/**
 * Pluggable api router if you're using Symfony and the FosJsRouting Bundle
 *
 * @see https://github.com/isychev/fos-routing/
 * @see https://github.com/FriendsOfSymfony/FOSJsRoutingBundle
 */
export class FosJsRoutingRouter extends Router {
  constructor (fosRouter, options = {}) {
    super()

    this.fosRouter = fosRouter
    this.routeExtractionOptions = Object.assign({
      prefix: '',
      resourceNamePattern: '[a-b]+',
      routePatterns: {
        resourceRoute: '{prefix}_{resourceName}_{action}'
      },
      actionPatterns: {
        list: 'list',
        get: 'get',
        create: 'create',
        replace: 'replace',
        update: 'update',
        delete: 'delete'
      }
    }, options)
  }

  async updateRoutes () {
    // TODO: extract routes from fosRouter based on naming scheme or something
    this.fosRouter.map(this.fosRouter.getRoutes(), name => {
      if (this.matchRouteName(name)) {
        const route = Route.fromPojo()
        this.addRoute(route)
      }
    })
  }

  matchRouteName (name) {

  }

  prepareResourceMatchRegex () {
    const pattern = this.routeExtractionOptions.routePatterns.resourceRoute
      .replace(
        '{prefix}',
        this.routeExtractionOptions.prefix
      )
      .replace(
        '{resourceName}',
        `(?<resourceName>${this.routeExtractionOptions.resourceNamePattern})`
      )
      .replace(
        '{action}',
        `(?<action>${Object.values(this.routeExtractionOptions.actionPatterns).join('|')})`
      )

    return new RegExp(pattern)
  }
}
