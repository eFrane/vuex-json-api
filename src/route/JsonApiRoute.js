import { Route } from './Route'

function getRequiredAttribute (routeResource, attributeName) {
  if (!Object.prototype.hasOwnProperty.call(routeResource.attributes, attributeName)) {
    throw new Error(`Missing required resource attribute: ${attributeName}`)
  }

  return routeResource.attributes[attributeName]
}

export class JsonApiRoute extends Route {
  /**
   * Create a route from a Json:Api resource of the VuexJsonApiRoute-type:
   *
   * ``` json
   * {
   *   "type": "VuexJsonApiRoute",
   *   "attributes": {
   *     "module": "themodule",
   *     "method": "thehttpmethod"
   *     "url": "The URL of the Resource, can be relative to baseURL",
   *     "parameters": ["Parameters", "for", "the", "above", url"]
   *   }
   * }
   * ```
   *
   * @param routeResource
   */
  constructor (routeResource) {
    if (!Object.prototype.hasOwnProperty.call(routeResource, 'attributes') ||
      !Object.prototype.hasOwnProperty.call(routeResource, 'type')) {
      throw new Error('Expected json:api-like object structure')
    }

    if (routeResource.type.toLowerCase() !== 'vuexjsonapiroute') {
      throw new Error('Invalid resource type')
    }

    const module = getRequiredAttribute(routeResource, 'module')
    const method = getRequiredAttribute(routeResource, 'method')
    const url = getRequiredAttribute(routeResource, 'url')
    const parameters = routeResource.attributes.parameters || []

    super(module, method, url, parameters)
  }
}
