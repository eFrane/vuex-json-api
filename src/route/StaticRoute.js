import { Route } from './Route'

/**
 * Parses a static route data object into a `Route`
 *
 * ``` json
 * [
 *   {
 *     "module": "MyApiModule",
 *     "action": "GET",
 *     "url": "/api/my-api-module/{id}",
 *     "parameters": ["id"]
 *   }
 * ]
 * ```
 */
export class StaticRoute extends Route {
  constructor (routeData) {
    const module = routeData.module
    const action = routeData.action
    const url = routeData.url
    const parameters = routeData.parameters || []

    super(module, action, url, parameters)
  }
}
