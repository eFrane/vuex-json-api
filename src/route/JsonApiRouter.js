import { Api } from '../api/Api'
import { Router } from './Router'
import { JsonApiRoute } from './JsonApiRoute'

/**
 * A simple router implementation querying a json:api endpoint which delivers all
 * necessary route information
 *
 * The endpoint must return a list of Route objects:
 *
 * ```json
 * {
 *     "data": [
 *         {
 *             "type": "Route",
 *             "id": "api.route.list",
 *             "attributes": {
 *                 "parameters": [],
 *                 "url": "api/route",
 *                 "method": "list"
 *             }
 *        },
 *        {
 *            "type": "Route",
 *            "id": "api.route.get",
 *            "attributes": {
 *                "parameters": [
 *                    "id"
 *                ],
 *                "url": "api/route/{id}",
 *                "method": "get"
 *            }
 *        }
 *     ]
 * }
 * ```
 *
 * To initialize this router, simply `new` it with the desired fetch path,
 * e.g. `new JsonApiRouter('/api/route')`.
 */
export class JsonApiRouter extends Router {
  constructor (fetchPath) {
    super()
    this.fetchPath = fetchPath
  }

  /**
   * Update routes
   *
   * @returns {*}
   */
  async updateRoutes () {
    const api = new Api()
    return api.get(this.fetchPath)
      .then(({ data }) => {
        console.time('router_setup')
        for (const idx in data.route) {
          if (Object.prototype.hasOwnProperty.call(data.route, idx)) {
            const routeResource = data.route[idx]

            this.addRoute(new JsonApiRoute(routeResource))
          }
        }
        console.timeEnd('router_setup')

        // returning self to make working with this in chained promises easier
        return this
      })
  }
}
