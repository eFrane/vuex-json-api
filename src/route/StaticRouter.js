import { Router } from './Router'
import { StaticRoute } from './StaticRoute'

/**
 * Static router
 *
 * It may be desirable to not do an additional request
 * for getting the api routes and instead bake them into
 * the code. To this avail, the `StaticRouter` can be
 * initialized with a POJO of { id, url, parameters }
 * whereas parameters are parts of the url which can be replaced.
 *
 *
 */
export class StaticRouter extends Router {
  /**
   * Requires an array of route data objects.
   *
   * @see StaticRoute
   * @param {Array} routes
   */
  constructor (routes) {
    super()

    for (const idx in routes) {
      const routeData = routes[idx]
      this.addRoute(new StaticRoute(routeData))
    }
  }
}
