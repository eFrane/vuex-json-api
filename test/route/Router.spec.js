import { Route } from '@/route/Route'
import { Router } from '@/route/Router'

describe('Router', () => {
  it('has routes', () => {
    const r = new Router()

    expect(r.routes).toStrictEqual({})
  })

  it('adds routes to the expected keys', () => {
    const route = new Route('module', 'GET', '/', [])
    const router = new Router()

    expect(router.routes.module).toBeUndefined()
    router.addRoute(route)
    expect(router.routes.module).toMatchSnapshot()
  })
})
