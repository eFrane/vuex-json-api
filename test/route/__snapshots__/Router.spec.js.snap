// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router adds routes to the expected keys 1`] = `
Object {
  "get": Route {
    "action": "get",
    "module": "module",
    "parameters": Array [],
    "url": "/",
  },
}
`;
